lightning/src/routing/network_graph.rs

```diff
@@ -312,12 +320,100 @@ impl<C: Deref + Sync + Send, L: Deref + Sync + Send> RoutingMessageHandler for N
-		// TODO
-		Err(LightningError {
-			err: String::from("Not implemented"),
-			action: ErrorAction::IgnoreError,
-		})
+	/// Processes a query from a peer by finding channels whose funding UTXOs
+	/// are in the specified block range. Due to message size limits, large range
+	/// queries may result in several reply messages. This implementation enqueues
+	/// all reply messages into pending events.
+	fn handle_query_channel_range(&self, their_node_id: &PublicKey, msg: QueryChannelRange) -> Result<(), LightningError> {
+		log_debug!(self.logger, "Handling query_channel_range peer={}, first_blocknum={}, number_of_blocks={}", log_pubkey!(their_node_id), msg.first_blocknum, msg.number_of_blocks);
+
+		let network_graph = self.network_graph.read().unwrap();
+
+		let start_scid = scid_from_parts(msg.first_blocknum, 0, 0);
+
+		// We receive valid queries with end_blocknum that would overflow SCID conversion.
+		// Manually cap the ending block to avoid this overflow.
+		let exclusive_end_scid = scid_from_parts(cmp::min(msg.end_blocknum(), 0xffffff), 0, 0);
+
+		// Per spec, we must reply to a query. Send an empty message when things are invalid.
+		if msg.chain_hash != network_graph.genesis_hash || start_scid.is_err() || exclusive_end_scid.is_err() {
+			let mut pending_events = self.pending_events.lock().unwrap();
+			pending_events.push(MessageSendEvent::SendReplyChannelRange {
+				node_id: their_node_id.clone(),
+				msg: ReplyChannelRange {
+					chain_hash: msg.chain_hash.clone(),
+					first_blocknum: msg.first_blocknum,
+					number_of_blocks: msg.number_of_blocks,
+					sync_complete: true,
+					short_channel_ids: vec![],
+				}
+			});
+			return Ok(());
+		}
+
+		// Creates channel batches. We are not checking if the channel is routable
+		// (has at least one update). A peer may still want to know the channel
+		// exists even if its not yet routable.
+		let mut batches: Vec<Vec<u64>> = vec![Vec::with_capacity(self.max_reply_scids)];
+		for (_, ref chan) in network_graph.get_channels().range(start_scid.unwrap()..exclusive_end_scid.unwrap()) {
+			if let Some(chan_announcement) = &chan.announcement_message {
+				// Construct a new batch if last one is full
+				if batches.last().unwrap().len() == batches.last().unwrap().capacity() {
+					batches.push(Vec::with_capacity(self.max_reply_scids));
+				}
+
+				let batch = batches.last_mut().unwrap();
+				batch.push(chan_announcement.contents.short_channel_id);
+			}
+		}
+		drop(network_graph);
+
+		let mut pending_events = self.pending_events.lock().unwrap();
+		let mut batch_index = 0;
+		let batch_count = batches.len();
+		for batch in batches.into_iter() {
+			// Per spec, the initial first_blocknum needs to be <= the query's first_blocknum.
+			// Use the query's values since we don't use pre-processed reply ranges.
+			let first_blocknum = if batch_index == 0 {
+				msg.first_blocknum
```

Comment:
nit: cant we just cheat and always send `msg.first_blocknum`? There's nothing that says we cant send 0, and a receiver needs to parse individual values to figure out the real block anyway.

Reply:
I'd like to. Legacy LND replied with the query's `first_blocknum` and `number_of_blocks` using `sync_complete` field to track whether a query was fully replied to, much cleaner that way.  Most implementations recognized this as valid, not sure if they still do. Then LND changed to use implicit sequencing along with with c-lightning and Eclair.  Now we're back to using `sync_complete` to drive sequencing but are still maintaining the sequencing for backwards compatibility reasons. So, long story, sequencing is probably required unless we lobby for a spec change or accept some nodes may not like it.

Reply:
Hmm, but my point was the spec doesn't actually seem to require it? Nodes might, but my pseudo-careful read of the spec seemed to say you can send the query params in every message.

Reply:
Right, but equal or greater can be equal :).

Reply:
Doh, yes it is! Some day I'll learn logic haha. Will clean this up!